proposals/0003-external-calls.md

```diff
@@ -0,0 +1,344 @@
+
+## Proposed solution
+The following solution is partially based upon the [Command Design Pattern]() and the [Oraclize Engine](https://docs.oraclize.it/). They allow for execution of the argument if other given conditions are met (as specified by the compiler). A valid external call should specify the following, some of these can be auto-filled by the compiler:
+- The contract address
+- The function name
+- The parameters
+- The gas allocation
+- The ether allocation
+
+Our motivations are:
+1. Contracts are untrustworthy by default;
+1. Arbitrary code may get executed;
+1. Failures may be silent;
+1. Interfaces may be incorrectly specified.
+
+We separate external calls into two types: _Trusted Calls_ and _Distrusted Calls_. Trusted calls are those accessed through the Flint Package Manager  or those which Flint has the source files for and deploys internally to the contract, i.e. Hub and Spoke Topology. Distrusted calls are those with an ABI interface or Trait interface.
+
+Distrusted calls (1) should be visually flagged in the source language as dangerous. Using a bang (!) would be consistent with the attempt call syntax for forcing a call without all information. In order to make a call we should specify the parameters for the call and to provide flexibility the default parameters should be at their minimum values. For instance the default gas provided should be 2300 (the amount given for just sending ether) with an option to send all gas.
+
+Educated calls meanwhile are not guaranteed to not introduce errors, but they have certain guarantees attached. These means that it combats (1), (2), (4):
+1. That there is a defined function at the end of the call
+2. That function obeys the modifiers given
+3. That function has the same return types and parameter types as defined
+4. The contract you call matches the source code given
+5. The gas provided should be inferred by gas estimation over flint
+
+We propose a method to both declare this interface within Flint, use the The Flint Package Manager to extract an interface, or call contracts distrusted.
+
+In _Distrusted Calls_, the contract (Client), sets the properties of the director which then sets up the command which is finally sent to the contract.
+The aim is to encapsulate a request as an object, thereby letting Flint parametrize clients with different requests.
+
+(3) is combated by having necessary catching of all external calls - or prefixing with `try!` and `try?` to revert and nullify respectively.
+
+### Distrusted Calls
+#### Interface specified
+```swift
+interface Alpha(State1, State2) {
+  var owner: Address
+
+  Alpha @(State1) :: (any) {
+    func doesNothing()
+  }
+  Alpha @(any) :: (owner) {
+    func doesNothingWithArgs(Int, Int, Int)
+    func withdraw() -> Int
+    func deposit(Int) -> Bool
+  }
+  @payable
+  func expensiveFunction()
+}
+
+let alpha: Director<Alpha> = 0x000... with Alpha
+
+try alpha!.doesNothing() {
```

Comment:
I feel like these try blocks would be clearer with another keyword before the open brace, e.g:

    try alpha!.doesNothing() then {